src/async-wrap.cc

```diff
@@ -407,6 +407,47 @@ static void DisablePromiseHook(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+class DestroyParam {
+ public:
+  double asyncId;
+  v8::Persistent<Object> target;
+  v8::Persistent<Object> propBag;
+};
+
+
+void AsyncWrap::WeakCallback(
+  const v8::WeakCallbackInfo<DestroyParam>& info
+) {
+  HandleScope scope(info.GetIsolate());
+
+  Environment* env = Environment::GetCurrent(info.GetIsolate());
+  DestroyParam* p = info.GetParameter();
+  Local<Object> propBag = PersistentToLocal(info.GetIsolate(), p->propBag);
+
+  Local<Value> val = propBag->Get(env->destroyed_string());
+  if (!val->BooleanValue()) {
+    PushBackDestroyAsyncId(env, p->asyncId);
+  }
+  p->propBag.Reset();
+  delete p;
+}
+
+
+static void RegisterDestroyHook(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsNumber());
+  CHECK(args[2]->IsObject());
+
+  Isolate* isolate = args.GetIsolate();
+  DestroyParam* p = new DestroyParam();
+  p->asyncId = args[1]->NumberValue();
```

Comment:
Can you use `args[1].As<Number>()->Value()`? This is deprecated (and slower for when we don’t have `val->IsNumber() == true`).